### PR TITLE
Faster flatten for some reports

### PR DESCRIPTION
**Pull request by tsteur**
refs #5098
- Makes flatten of Actions.*, Events.getCategory, Events.getAction, Events.getName, Referrers.getWebsites much faster and if it failed before it should not anymore
- Many flatten operations did not work correctly before and either not flatten or did it wrong
- Makes all other kind of normal reports faster as well
- Many bugfixes for flatten and for other reports (eg Reports Total Calculation)

I will rebase and squash before merging, please let me know!

### CODE REVIEW

File: core/API/DataTableManipulator/Flattener.php [hunk 1]
```diff
@@ -38,17 +38,16 @@ public function includeAggregateRows()
...
-    public $recursiveLabelSeparator = ' - ';
+    public $recursiveLabelSeparator = '';
 
     /**
      * @param  DataTable $dataTable
+     * @param string $recursiveLabelSeparator
      * @return DataTable|DataTable\Map
      */
-    public function flatten($dataTable)
+    public function flatten($dataTable, $recursiveLabelSeparator)
```
Comment (tsteur) [inline core/API/DataTableManipulator/Flattener.php]: $recursiveLabelSeparator is no longer hard coded / hacked in the Flattener, it can be defined by each report now.

File: core/API/DataTableManipulator/Flattener.php [hunk 2]
```diff
@@ -84,15 +84,21 @@ protected function manipulateDataTable($dataTable)
...
         if ($label !== false) {
             $label = trim($label);
-            if (substr($label, 0, 1) == '/' && $this->recursiveLabelSeparator == '/') {
-                $label = substr($label, 1);
+
+            if ($this->recursiveLabelSeparator == '/') {
+                if (substr($label, 0, 1) == '/') {
+                    $label = substr($label, 1);
+                } elseif ($rowId === DataTable::ID_SUMMARY_ROW && $labelPrefix && $label != DataTable::LABEL_SUMMARY_ROW) {
+                    $label = ' - ' . $label;
```
Comment (tsteur) [inline core/API/DataTableManipulator/Flattener.php]: We did not have any tests for this, I made the output nicer eg instead of `/pa/th/Others` it is now `/pa/th - Others` as "Others" could be a normal folder / file

File: core/API/DataTableManipulator/ReportTotalsCalculator.php [hunk 3]
```diff
@@ -66,14 +66,19 @@ protected function manipulateDataTable($dataTable)
         $firstLevelTable    = $this->makeSureToWorkOnFirstLevelDataTable($dataTable);
         $metricsToCalculate = Metrics::getMetricIdsToProcessReportTotal();
 
+        $realMetricNames = array();
```
Comment (tsteur) [inline core/API/DataTableManipulator/ReportTotalsCalculator.php]: This is all done to make it faster, eg do not get readable column name as often. This here is executed quite often

File: core/API/DataTablePostProcessor.php [hunk 4]
```diff
@@ -145,7 +145,13 @@ public function applyFlattener($dataTable)
             if (Common::getRequestVar('include_aggregate_rows', '0', 'string', $this->request) == '1') {
                 $flattener->includeAggregateRows();
             }
-            $dataTable = $flattener->flatten($dataTable);
+
+            $recursiveLabelSeparator = ' - ';
+            if ($this->report) {
+                $recursiveLabelSeparator = $this->report->getRecursiveLabelSeparator();
```
Comment (tsteur) [inline core/API/DataTablePostProcessor.php]: As mentioned above the separator is now configured by a report

File: core/Archive.php [hunk 5]
```diff
@@ -474,6 +475,38 @@ public static function getDataTableFromArchive($name, $idSite, $period, $date, $
...
+     * @param string $period @see {@link build()}
+     * @param string $date @see {@link build()}
+     * @param string $segment @see {@link build()}
+     * @param bool $expanded If true, loads all subtables. See {@link getDataTableExpanded()}
+     * @param bool $flat If true, loads all subtables and disabled all recursive filters.
+     * @param int|null $idSubtable See {@link getDataTableExpanded()}
+     * @param int|null $depth See {@link getDataTableExpanded()}
+     * @return DataTable|DataTable\Map
+     */
+    public static function createDataTableFromArchive($recordName, $idSite, $period, $date, $segment, $expanded = false, $flat = false, $idSubtable = null, $depth = null)
```
Comment (tsteur) [inline core/Archive.php]: I introduced a new method here since I did not want to change the behaviour of the existing `getDataTableFromArchive`

File: core/DataTable.php [hunk 6]
```diff
@@ -342,6 +342,17 @@ public function __destruct()
     }
 
     /**
+     * @ignore
+     * does not update the summary row!
+     */
+    public function setRows($rows)
```
Comment (tsteur) [inline core/DataTable.php]: All the changes here and the new methods are needed for faster sorting in the `Sort` filter

File: core/DataTable/Filter/Sort.php [hunk 7]
```diff
@@ -21,7 +21,8 @@
  *
  * @api
  */
-class Sort extends BaseFilter
+class
+Sort extends BaseFilter
```
Comment (tsteur) [inline core/DataTable/Filter/Sort.php]: faster sort is big win!

File: core/DataTable/Filter/Sort.php [hunk 8]
```diff
@@ -243,6 +231,51 @@ public function filter($table)
...
+    private function sort(DataTable $table, $functionCallback)
+    {
+        $table->setTableSortedBy($this->columnToSort);
+
+        $rows = $table->getRowsWithoutSummaryRow();
+
+        // get column value and label only once for performance tweak
+        $values = array();
+        foreach ($rows as $key => $row) {
+            $values[$key] = array($this->getColumnValue($row), $row->getColumn('label'));
```
Comment (tsteur) [inline core/DataTable/Filter/Sort.php]: On 25k rows `getColumnValue` was executed > 1 mio times. Now we do it only 25k times which brings big performance boost

File: core/DataTable/Filter/Truncate.php [hunk 9]
```diff
@@ -96,7 +96,7 @@ private function addSummaryRow($table)
             return;
         }
 
-        $rows   = $table->getRows();
+        $rows   = array_values($table->getRows());
```
Comment (tsteur) [inline core/DataTable/Filter/Truncate.php]: Truncate was buggy before if some rows were deleted and was summarizing wrong values under circumstances since it always reused the summary row when an array index was missing

File: core/DataTable/Row.php [hunk 10]
```diff
@@ -209,37 +209,9 @@ public function getColumn($name)
             return false;
         }
 
-        if ($this->isColumnValueCallable($this->c[self::COLUMNS][$name])) {
```
Comment (tsteur) [inline core/DataTable/Row.php]: Callable column values are no longer used since the processed metrics refactoring, big performance win

File: core/Plugin/API.php [hunk 11]
```diff
@@ -41,5 +41,5 @@
  */
 abstract class API extends Singleton
 {
-
+
```
Comment (tsteur) [inline core/Plugin/API.php]: I could undo this change if anyone is annoyed by that, I don't care :)

File: core/Plugin/Metric.php [hunk 12]
```diff
@@ -108,26 +108,41 @@ public function beforeFormat($report, DataTable $table)
...
-        }
-
         if ($row instanceof Row) {
             $value = $row->getColumn($columnName);
-            if ($value === false
-                && isset($mappingNameToId[$columnName])
-            ) {
-                $value = $row->getColumn($mappingNameToId[$columnName]);
+
+            if ($value === false) {
```
Comment (tsteur) [inline core/Plugin/Metric.php]: It does actually not often happen that we do not have a value therefore get the MappingFromNameToId only if needed => performance win

File: core/UrlHelper.php [hunk 13]
```diff
@@ -231,7 +231,10 @@ public static function getPathAndQueryFromUrl($url)
         $parsedUrl = parse_url($url);
         $result = '';
         if (isset($parsedUrl['path'])) {
-            $result .= substr($parsedUrl['path'], 1);
+            if (substr($parsedUrl['path'], 0, 1) == '/') {
+                $parsedUrl['path'] = substr($parsedUrl['path'], 1);
```
Comment (tsteur) [inline core/UrlHelper.php]: There was a bug if `$url` was eg `Others` it resulted before in `thers` see test.

File: plugins/Actions/API.php [hunk 14]
```diff
@@ -445,28 +452,15 @@ protected function doFilterPageDatatableSearch($callBackParameters, $table, $sea
...
      *
      * @param DataTable|DataTable\Simple|DataTable\Map $dataTable
-     * @param bool $expanded
      */
-    protected function filterActionsDataTable($dataTable, $expanded = false)
+    private function filterActionsDataTable($dataTable)
     {
         // Must be applied before Sort in this case, since the DataTable can contain both int and strings indexes
         // (in the transition period between pre 1.2 and post 1.2 datatable structure)
-        $dataTable->filter('ReplaceColumnNames');
```
Comment (tsteur) [inline plugins/Actions/API.php]: This all moved into a separate filter so it is better reusable and would be also better testable

File: plugins/Events/API.php [hunk 15]
```diff
@@ -213,20 +217,4 @@ public function getCategoryFromNameId($idSite, $period, $date, $idSubtable, $seg
     {
         return $this->getDataTable(__FUNCTION__, $idSite, $period, $date, $segment, $expanded = false, $idSubtable);
     }
-
-    /**
-     * @param DataTable $dataTable
-     */
-    protected function filterDataTable($dataTable)
-    {
-        $dataTable->filter('Sort', array(Metrics::INDEX_NB_VISITS));
```
Comment (tsteur) [inline plugins/Events/API.php]: * Sort was not needed => Big performance win
* The other two queued filters were already assigned and not needed at all

File: tests/PHPUnit/System/expected/test_CustomEvents_Events.getAction_flat__API.getProcessedReport_day.xml [hunk 16]
```diff
@@ -235,7 +231,7 @@
 		<result prettyDate="Saturday 9 January 2010" />
 	</reportMetadata>
 	<reportTotal>
-		<nb_visits>38</nb_visits>
-		<nb_uniq_visitors>38</nb_uniq_visitors>
+		<nb_visits>42</nb_visits>
```
Comment (tsteur) [inline tests/PHPUnit/System/expected/test_CustomEvents_Events.getAction_flat__API.getProcessedReport_day.xml]: This is a fix as ReportsTotalsCalculator does no longer operate on the flattened table but only on the first level of the report which is more accurate

File: tests/PHPUnit/System/expected/test_FlattenReports_withAggregate__Actions.getPageUrls_week.xml [hunk 17]
```diff
@@ -387,6 +387,9 @@
 		<nb_visits>1</nb_visits>
 		<nb_hits>1</nb_hits>
 		<sum_time_spent>0</sum_time_spent>
+		<nb_hits_with_time_generation>0</nb_hits_with_time_generation>
```
Comment (tsteur) [inline tests/PHPUnit/System/expected/test_FlattenReports_withAggregate__Actions.getPageUrls_week.xml]: I don't know why it was not present before but it seems to be a fix

**Comment by mnapoli** [comment]
Is it a collection of small fixes and improvements or one big refactoring? If it's one big change, it would be interesting to share an overview of what you did because reading the diff is not really easy. Also do you have any numbers (even rough ones) to share regarding the perf improvements?

**Comment by tsteur** [comment]
It is a collection of small fixes and improvements see comments. No big refactoring. I only made many things faster like sorting etc see comments.

One big performance improvement was to no longer request each subtable separately but load the expanded datatable at once and then flatten the expanded table.

It is hard to show any numbers since in most cases flatten crashed after X seconds/minutes and it was often not even possible to flatten. Not even on demo etc which now loads in < 1 second. Also it depends on the structure of the datatable and of the report. Eg whether there are many first level rows, how many subtables the tables has etc. Basically one can say a flattened table with 25k rows in the UI takes less than 5 seconds, in an API call maybe in even 2 or 3 seconds. On a fast CPU maybe even < 1 or 2 seconds. The more rows a datatable has the better the benefit will be.

File: plugins/Actions/DataTable/Filter/Actions.php [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+        $table->filter(function (DataTable $dataTable) {
+            foreach ($dataTable->getRows() as $row) {
+                $url = $row->getMetadata('url');
+                if ($url) {
+                    $row->setMetadata('segmentValue', urldecode($url));
+                }
+            }
+        });
+
+        // TODO can we remove this one again?
```
Comment (mattab) [inline plugins/Actions/DataTable/Filter/Actions.php]: Not sure if this TODO should still be there?

File: plugins/Referrers/API.php [hunk 19]
```diff
@@ -296,22 +295,24 @@ public function getKeywordsFromCampaignId($idSite, $period, $date, $idSubtable,
...
     public function getUrlsFromWebsiteId($idSite, $period, $date, $idSubtable, $segment = false)
     {
         $dataTable = $this->getDataTable(Archiver::WEBSITES_RECORD_NAME, $idSite, $period, $date, $segment, $expanded = false, $idSubtable);
-        // the htmlspecialchars_decode call is for BC for before 1.1
-        // as the Referrer URL was previously encoded in the log tables, but is now recorded raw
-        $dataTable->queueFilter('ColumnCallbackAddMetadata', array('label', 'url', function ($label) {
-            return htmlspecialchars_decode($label);
-        }));
-        $dataTable->queueFilter('ColumnCallbackReplace', array('label', __NAMESPACE__ . '\getPathFromUrl'));
+        $dataTable->filter('Piwik\Plugins\Referrers\DataTable\Filter\UrlsFromWebsiteId');
```
Comment (mattab) [inline plugins/Referrers/API.php]: the filter used to be queued and is now directly executed - is it on purpose? if not, maybe queuing is faster as we'd only filter on the eg. 100 displayed rows VS possibly many more rows

File: plugins/CoreHome/Columns/Metrics/CallableProcessedMetric.php [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+    public function compute(Row $row)
+    {
+        if ($this->callback) {
+            return call_user_func($this->callback, $row);
+        }
+    }
+
+    public function getTranslatedName()
+    {
+        return '';
```
Comment (mattab) [inline plugins/CoreHome/Columns/Metrics/CallableProcessedMetric.php]: maybe this should throw an exception instead?

**Comment by mattab** [comment]
I'll merge it so we can test on the demo. Lots of small CPU friendly changes :+1:

File: plugins/Actions/DataTable/Filter/Actions.php (hunk 18 above)
Comment (tsteur) [inline plugins/Actions/DataTable/Filter/Actions.php]: Yes, it should be there. We need to discuss this end of week

File: plugins/Referrers/API.php (hunk 19 above)
Comment (tsteur) [inline plugins/Referrers/API.php]: It has to be executed directly as this filter queues the other filters

**Comment by mattab** [comment]
Some UI tests fail because of secondary column sorting: http://builds-artifacts.piwik.org/ui-tests.master/10794.7/screenshot-diffs/diffviewer.html
eg. in http://builds-artifacts.piwik.org/ui-tests.master/10794.7/screenshot-diffs/singlediff.html?processed=../processed-ui-screenshots/UIIntegrationTest_ecommerce_sales.png&expected=UIIntegrationTest_ecommerce_sales.png&github=UIIntegrationTest_ecommerce_sales.png
I reckon we should sort by nb_visits as secondary column sort, in all reports where nb_visits is defined, I think it's what users will expect

**Comment by tsteur** [comment]
The UI tests fail because of the other pull request as mentioned there. It is not a bug

**Comment by mattab** [comment]
the secondary column sort request is followed up in #7401
